sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,362 @@ object AggregateUtils {
   }
 }
 
-abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
-    plan: INPUT,
-    aggRequiredChildDistributionExpressions: Option[Seq[Expression]],
-    conf: RapidsConf,
-    parent: Option[RapidsMeta[_, _, _]],
-    rule: DataFromReplacementRule) extends SparkPlanMeta[INPUT](plan, conf, parent, rule) {
-
-  val agg: BaseAggregateExec
-
-  private val requiredChildDistributionExpressions: Option[Seq[BaseExprMeta[_]]] =
-    aggRequiredChildDistributionExpressions.map(_.map(GpuOverrides.wrapExpr(_, conf, Some(this))))
-  private val groupingExpressions: Seq[BaseExprMeta[_]] =
-    agg.groupingExpressions.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
-  private val aggregateExpressions: Seq[BaseExprMeta[_]] =
-    agg.aggregateExpressions.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
-  private val aggregateAttributes: Seq[BaseExprMeta[_]] =
-    agg.aggregateAttributes.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
-  private val resultExpressions: Seq[BaseExprMeta[_]] =
-    agg.resultExpressions.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
-
-  override val childExprs: Seq[BaseExprMeta[_]] =
-    requiredChildDistributionExpressions.getOrElse(Seq.empty) ++
-        groupingExpressions ++
-        aggregateExpressions ++
-        aggregateAttributes ++
-        resultExpressions
-
-  override def tagPlanForGpu(): Unit = {
-    agg.groupingExpressions
-      .find(_.dataType match {
-        case _@(ArrayType(_, _) | MapType(_, _, _)) | _@StructType(_) => true
-        case _ => false
-      })
-      .foreach(_ =>
-        willNotWorkOnGpu("Nested types in grouping expressions are not supported"))
-    if (agg.resultExpressions.isEmpty) {
-      willNotWorkOnGpu("result expressions is empty")
-    }
-
-    tagForReplaceMode()
-
-    if (agg.aggregateExpressions.exists(expr => expr.isDistinct)
-        && agg.aggregateExpressions.exists(expr => expr.filter.isDefined)) {
-      // Distinct with Filter is not supported on the GPU currently,
-      // This makes sure that if we end up here, the plan falls back to the CPU
-      // which will do the right thing.
-      willNotWorkOnGpu(
-        "DISTINCT and FILTER cannot be used in aggregate functions at the same time")
+/** Utility class to hold all of the metrics related to hash aggregation */
+case class GpuHashAggregateMetrics(
+    numOutputRows: GpuMetric,
+    numOutputBatches: GpuMetric,
+    totalTime: GpuMetric,
+    computeAggTime: GpuMetric,
+    concatTime: GpuMetric,
+    sortTime: GpuMetric,
+    spillCallback: RapidsBuffer.SpillCallback)
+
+/** Utility class to convey information on the aggregation modes being used */
+case class AggregateModeInfo(
+    uniqueModes: Seq[AggregateMode],
+    hasPartialMode: Boolean,
+    hasFinalMode: Boolean,
+    hasCompleteMode: Boolean)
+
+class GpuHashAggregateIterator(
```

Comment:
Could we get a high level comment on this class, or somewhere in this file that documents the data flow? I understand it now, but it took me two hours of reading the code and jumping back and forth to wrap my head around it. And that is even after we had a call last week and came up with a design for this together.  Just a quick note like. 

The data goes through a few stages.  The first stage does an initial aggregation pass across all of the input batches `aggregateInputBatches`. At that point it needs to either combine all of the batches together into a single output batch `mergeAggregatedBatches` or fall back to sorting this intermediate data so the results can be combined and output as multiple batches `buildSortFallbackIterator`.